### PR TITLE
[logstash bridge]: bootstrap

### PR DESCRIPTION
**Pull request by yaauie**
Adds a new `logstash-bridge` project in `/libs` that:
 - publishes bridge methods for accessing various constructors, static functions, and static fields that are required by Logstash's Elastic Integration Filter (these are _not_ a part of the distributed artifact; the LS plugin will build and use directly)
 - adds "duck type" interface validation tests to ensure that the object methods that Logstash's Elastic Integration Filter rely on maintain a stable API.

## TODO:

 - [ ] hook `:libs:elasticsearch-logstash-bridge:test` into the build/CI, so that when someone breaks an API Logstash relies on we have direct feedback.
 - [ ] map codeowners to a public team

### CODE REVIEW

**Comment by mark-vieira** [comment]
> * publishes bridge methods for accessing various constructors, static functions, and static fields that are required by Logstash's Elastic Integration Filter (these are _not_ a part of the distributed artifact; the LS plugin will build and use directly)

Is this a new or existing plugin? Is that plugin going to be part of this repository or do we need to publish this artifact somewhere for external consumption?

**Comment by yaauie** [comment]
> Is this a new or existing plugin? Is that plugin going to be part of this repository or do we need to publish this artifact somewhere for external consumption?

The [elastic/logstash-filter-elastic_integration](https://github.com/elastic/logstash-filter-elastic_integration) exists, and is a plugin for _Logstash_, whose artifacts are published as self-contained rubygems (_no_ maven). The plugin already builds and bundles a minimal subset of Elasticsearch as part of its build process, and is capable of also building this artifact for self-consumption without requiring it to be published elsewhere.

File: libs/logstash-bridge/build.gradle [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+apply plugin: 'elasticsearch.build'
+
+dependencies {
+  compileOnly project(':server')
```
Comment (mark-vieira) [inline libs/logstash-bridge/build.gradle:11]: @rjernst I vaguely remember us not wanting libs to depend on server but this isn't really a "lib" of traditional sorts.

Comment (yaauie) [reply libs/logstash-bridge/build.gradle:11]: I'm glad to place it elsewhere, even in x-pack project as this does not need to be published under an OSS license. Discussions with @rjernst offline indicated that `/libs` was an appropriate starting-point.

> [The] way i would do this is have a lib, it will depend on server and painless. That lib will only live inside ES source, you can build/pull in the jars within logstash
> -- @rjernst via Slack

Comment (rjernst) [reply libs/logstash-bridge/build.gradle:11]: This is a tricky and special case. I think we are ok to depend on server here.

File: libs/logstash-bridge/build.gradle [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  compileOnly project(':modules:lang-painless')
+  compileOnly project(':modules:lang-painless:spi')
+
+  testImplementation(project(":test:framework")) {
+    exclude group: 'org.elasticsearch', module: 'logstashbridge'
```
Comment (mark-vieira) [inline libs/logstash-bridge/build.gradle:16]: Why is this required? Nothing should actually depend on this project so there should be no reason to exclude it.

File: libs/logstash-bridge/src/test/java/org/elasticsearch/logstashbridge/ThreadPoolBridgeTests.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.logstashbridge.DucktypeMatchers.instancesQuackLike;
+import static org.elasticsearch.logstashbridge.DucktypeMatchers.staticallyQuacksLike;
+
+public class ThreadPoolBridgeTests extends ESTestCase {
+
+    public void testThreadPoolAPIStability() throws Exception {
```
Comment (rjernst) [inline libs/logstash-bridge/src/test/java/org/elasticsearch/logstashbridge/ThreadPoolBridgeTests.java:26]: I don't understand the purpose of these tests. The point of this lib is that it is a bridge that will maintain all compatibility promises. If you need access to ThreadPool, have your own ThreadPool wrapper, and delegate in this bridge lib to the ES ThreadPool. By doing that, any refactorings (eg renamings/adjustments/whatever) of ThreadPool methods will require ES devs to update this code before merging, no notification to logstash devs necessary.

File: libs/logstash-bridge/build.gradle (hunk 2 above)
Comment (yaauie) [reply libs/logstash-bridge/build.gradle:16]: Looks lime a remnant from copying a different lib to bootstrap. I'll snip.

File: libs/logstash-bridge/src/test/java/org/elasticsearch/logstashbridge/ThreadPoolBridgeTests.java (hunk 3 above)
Comment (yaauie) [reply libs/logstash-bridge/src/test/java/org/elasticsearch/logstashbridge/ThreadPoolBridgeTests.java:26]: Got it.

For many of these objects, we need to pass actual instances of these objects to ES-owned code, or to work with instances that are provided _by_ ES-owned code.\

I had hoped to avoid many short-lived wrappers for each object by applying some pressure to the stability of the API, but I understand now that we are not to expect that to work.

I'll convert this PR back to 'draft' and work up a wrapper pattern for each of the objects we need to work with.
